snuba/snapshots/postgres_snapshot.py

```diff
@@ -0,0 +1,127 @@
+from __future__ import annotations
+
+import jsonschema  # type: ignore
+import json
+import logging
+
+from contextlib import contextmanager
+from dataclasses import dataclass
+from os import path as os_path
+from typing import NewType, Generator, IO, Sequence
+
+from snuba.snapshots import SnapshotDescriptor, SnapshotId, TableConfig
+from snuba.snapshots import BulkLoadSource
+
+Xid = NewType("Xid", int)
+
+
+@dataclass(frozen=True)
+class PostgresSnapshotDescriptor(SnapshotDescriptor):
+    """
+    Provides the metadata for the loaded snapshot.
+    """
+    id: SnapshotId
+    xmin: Xid
+    xmax: Xid
+    xip_list: Sequence[Xid]
+    tables: Sequence[TableConfig]
+
+
+logger = logging.getLogger('postgres-snapshot')
+
+
+class PostgresSnapshot(BulkLoadSource):
+    """
+    TODO: Make this a library to be reused outside of Snuba when after this
+    is validated in production.
+
+    Represents a snapshot from a Postgres instance and dumped into
+    a set of files (one per table).
+
+    This class know how to read and validate the dump.
+    """
+
+    def __init__(self, path: str, descriptor: PostgresSnapshotDescriptor) -> None:
+        self.__path = path
+        self.__descriptor = descriptor
+
+    @classmethod
+    def load(cls, path: str, product: str) -> PostgresSnapshot:
+        meta_file_name = os_path.join(path, "metadata.json")
+        with open(meta_file_name, "r") as meta_file:
+            json_desc = json.load(meta_file)
+            jsonschema.validate(
+                json_desc,
+                {
```

Comment:
It'd probably be a bit clearer to define this schema at module scope than inlining here.